### PR TITLE
SQL Task. Punko Pavel

### PR DESCRIPTION
**Pull request by PashaPunko**
https://travis-ci.org/github/PashaPunko/db-assignments/builds/719365909

### CODE REVIEW

File: task/sql-tasks.js [hunk 1]
```diff
@@ -105,7 +153,20 @@ async function task_1_6(db) {
  *
  */
 async function task_1_7(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            e.EmployeeID AS EmployeeId,
+            CONCAT(e.FirstName, ' ', e.LastName) AS FullName,
+            IFNULL((
+                SELECT
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use join

File: task/sql-tasks.js [hunk 2]
```diff
@@ -129,7 +203,14 @@ async function task_1_8(db) {
  *
  */
 async function task_1_9(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           CustomerID,
+           ContactName
+        FROM Customers
+        WHERE ContactName REGEXP '^F..n.*'
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use like

File: task/sql-tasks.js [hunk 3]
```diff
@@ -177,7 +288,18 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
```
Comment (AlexMokin) [inline task/sql-tasks.js:290]: it's better to write this query like this
notice that from-clause is optional.



```
select
   (select count()...) as 'T'
   (select count()...) as 'TT'
```

File: task/sql-tasks.js [hunk 4]
```diff
@@ -199,7 +329,23 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            SUM( IF( MONTH(OrderDate)=1 AND YEAR(OrderDate)=1997, 1, 0) ) AS January,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: check year in where clause

File: task/sql-tasks.js [hunk 5]
```diff
@@ -199,7 +329,23 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            SUM( IF( MONTH(OrderDate)=1 AND YEAR(OrderDate)=1997, 1, 0) ) AS January,
+            SUM( IF( MONTH(OrderDate)=2 AND YEAR(OrderDate)=1997, 1, 0) ) AS February,
+            SUM( IF( MONTH(OrderDate)=3 AND YEAR(OrderDate)=1997, 1, 0) ) AS March,
+            SUM( IF( MONTH(OrderDate)=4 AND YEAR(OrderDate)=1997, 1, 0) ) AS April,
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count records, it shows intent better

File: task/sql-tasks.js [hunk 6]
```diff
@@ -223,7 +377,21 @@ async function task_1_16(db) {
...
+           Categories.CategoryName AS CategoryName,
+           products.Price AS AvgPrice
+        FROM (
+            SELECT
+                CategoryID,
+                AVG(UnitPrice) AS Price
+            FROM Products
+            GROUP BY CategoryID
+            ) products
+            RIGHT JOIN Categories ON products.CategoryID=Categories.CategoryID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not use right join if not necessary

File: task/sql-tasks.js [hunk 7]
```diff
@@ -247,7 +424,30 @@ async function task_1_18(db) {
  *
  */
 async function task_1_19(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+    SELECT
```
Comment (AlexMokin) [inline task/sql-tasks.js:419]: implement without subquerires

File: task/sql-tasks.js [hunk 8]
```diff
@@ -259,7 +459,30 @@ async function task_1_19(db) {
  *
  */
 async function task_1_20(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
```
Comment (AlexMokin) [inline task/sql-tasks.js:442]: implement without subquerires

File: task/sql-tasks.js [hunk 9]
```diff
@@ -280,8 +513,41 @@ async function task_1_21(db) {
  * @return {array}
  */
 async function task_1_22(db) {
-    throw new Error("Not implemented");
-}
+        let result = await db.query(`
+            SELECT DISTINCT
```
Comment (AlexMokin) [inline task/sql-tasks.js:486]: implement with single subquery

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by PashaPunko** [comment]
https://travis-ci.org/github/PashaPunko/db-assignments/builds/722996535

File: task/sql-tasks.js [hunk 10]
```diff
@@ -129,7 +202,14 @@ async function task_1_8(db) {
  *
  */
 async function task_1_9(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           CustomerID,
+           ContactName
+        FROM Customers
+        WHERE ContactName LIKE 'F%n%'
```
Comment (AlexMokin) [inline task/sql-tasks.js]: Exactly 4th character must be n

File: task/sql-tasks.js [hunk 11]
```diff
@@ -223,7 +373,21 @@ async function task_1_16(db) {
  *
  */
 async function task_1_17(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+           Categories.CategoryName AS CategoryName,
+           products.Price AS AvgPrice
+        FROM (
+            SELECT
```
Comment (AlexMokin) [inline task/sql-tasks.js]: try to do it without a sub-query

**Comment by AlexMokin** [comment]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by PashaPunko** [comment]
https://travis-ci.org/github/PashaPunko/db-assignments/builds/723340423

**Comment by AlexMokin** [comment]
Accepted

**Pull request closed without merging**